### DISCUSSION TITLE
How to get observations (Pose / velocity of geoms) in the Frame of reference of an agent?

### DISCUSSION THREAD

**Discussion by shantnavagarwal** [google-deepmind/mujoco#2290] [Asking for Help] [answered]
### Intro

Hi!

I am a graduate student at TU Delft, I use MuJoCo for my research on Multi-agent Reinforcement Learning.

### My setup

Mujoco: 3.2.5
MJX: 3.2.5 (Not being used)
Python: 3.10

### My question

I have multiple agents within a single Mujoco environment. I am able to get values of any geom's pose and velocity in the world frame. I want to ask if there is a way to get these values in an agent's frame of reference. 

One approach I have found is to get the agent's frame quaternion and pos from mj_data and then use functions from the [Poses](https://mujoco.readthedocs.io/en/stable/APIreference/APIfunctions.html#poses) API. My question is: Is there  a direct way to specify a body (via it's ID) and get values from mj_data in the body's frame?

Thanks!

### Confirmations

- [x] I searched the [latest documentation](https://mujoco.readthedocs.io/en/latest/overview.html) thoroughly before posting.
- [X] I searched previous [Issues](https://github.com/google-deepmind/mujoco/issues) and [Discussions](https://github.com/google-deepmind/mujoco/discussions), I am certain this has not been raised before.

**Comment by yuvaltassa** [comment, marked as answer]
This is most easily accomplished with sensors. All "framexx" sensors accept a reference frame object.

**Reply by shantnavagarwal** [reply]
Thanks for giving the correct reference. [sensor/framepos](https://mujoco.readthedocs.io/en/stable/XMLreference.html#sensor-framepos)
